### ISSUE TITLE
Erroneous retrieval of derivation inputs for synchronous derivations

### ISSUE THREAD

**Issue by markushofmeister** [cambridge-cares/TheWorldAvatar#852]
There seems to be an issue when retrieving the key-value, i.e., type-instance, pairs for synchronous derivation inputs

When initialising a synchronous derivations like this

`
derivation_client.createSyncDerivationForNewInfo(agent.agentIRI, derivation_input_set, cf.ONTODERIVATION_DERIVATIONWITHTIMESERIES)
`,
the derivation_inputs.getInputs() method returns a dictionary with duplicated IRIs:
```
{
    "http://www.w3.org/2006/time#Duration": [
        "https://www.theworldavatar.com/test/Duration_1"
    ],
    "https://www.theworldavatar.com/kg/ontotimeseries/ForecastingModel": [
        "https://www.theworldavatar.com/test/ForecastingModel_1"
    ],
    "https://www.theworldavatar.com/kg/ontotimeseries/Frequency": [
        "https://www.theworldavatar.com/test/Frequency_1"
    ],
    "http://www.w3.org/2006/time#Interval": [
        "https://www.theworldavatar.com/test/OptimisationInterval_1"
    ],
	
    "https://www.theworldavatar.com/test/Duration_1": [
        "https://www.theworldavatar.com/test/Duration_1"
    ],
    "https://www.theworldavatar.com/test/OptimisationInterval_1": [
        "https://www.theworldavatar.com/test/OptimisationInterval_1"
    ],
    "https://www.theworldavatar.com/test/ForecastingModel_1": [
        "https://www.theworldavatar.com/test/ForecastingModel_1"
    ],
    "https://www.theworldavatar.com/test/Frequency_1": [
        "https://www.theworldavatar.com/test/Frequency_1"
    ]
}
```

When initialising the same `derivation_input_set` as asynchronous derivation using `createAsyncDerivationForNewInfo`, all derivation inputs are correctly retrieved.

**Comment by jb2197** [comment]
The identified issue is a result of an erroneous SPARQL query used when mapping the derivation inputs to the input signature of the agent. Specifically, below was the query actually performed before the fix:
```sparql
PREFIX derived: <https://www.theworldavatar.com/kg/ontoderivation/>
PREFIX agent: <http://www.theworldavatar.com/ontology/ontoagent/MSM.owl#>
SELECT DISTINCT ?input ?type
WHERE {
    VALUES (?input)  {  (<https://input1>)  (<https://input2>)  }
    {
        <https://agent> agent:hasOperation/agent:hasInput/agent:hasMandatoryPart/agent:hasType ?type .
    } UNION {
        # NOTE this part is empty, however it SHOULD NOT
    }
    ?input a*/<http://www.w3.org/2000/01/rdf-schema#subClassOf>* ?type .
}
```

The structure of the `UNION` in the query was introduced in #801 to support the situation where no outputs are produced from derivation. The second part of the `UNION` is added to accommodate the case when the upstream derivation doesn't have outputs when making up the downstream derivation. Such an issue was not observed in the past as the duplicated instances retrieved as inputs didn't affect the normal execution (no upstream derivations IRIs were missing so the framework functions), there were also no test cases on duplicate IRIs in the derivation inputs.

The missed part in the `UNION` is a `VALUES` clause: 
```sparql
VALUES ( ?type )  {  (derived:Derivation)  (derived:DerivationWithTimeSeries)  (derived:DerivationAsyn)  }
```
that should be generated from `derivationValuesPattern` in: https://github.com/cambridge-cares/TheWorldAvatar/blob/753bf69cf970325f96a5d6bdc2cfdc3a58c8a587/JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java#L1185

However, when it calls [`GraphPatterns::union`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/GraphPatterns.java#L175), the following codes are called by sequence:
[`return new GraphPatternNotTriples(union.union(patterns));`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/GraphPatterns.java#L178)
[`addElements(GraphPatterns::extractOrConvertToGGP, patterns);`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/AlternativeGraphPattern.java#L46)
[`return new GroupGraphPattern(pattern);`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/GraphPatterns.java#L249)
[`} else if (original != null && !original.isEmpty()) {`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/GroupGraphPattern.java#L46)

Note that when `original` is the `derivationValuesPattern` (object of `ValuesPattern` implemented in TWA, which extends `org.eclipse.rdf4j.sparqlbuilder.graphpattern.GraphPattern`), `original.isEmpty()` will return [the default value `true`](https://github.com/eclipse/rdf4j/blob/0afd64b3c6604950b9d67626582afc1fd1bc6f1d/core/sparqlbuilder/src/main/java/org/eclipse/rdf4j/sparqlbuilder/graphpattern/GraphPattern.java#L209). As a result, the VALUES clause is not added to the returned object of `GraphPatterns.union(agentTypePattern, derivationValuesPattern)`, consequently, not presented in the final SPARQL query string.

The open PR overrides the `ValuesPattern::isEmpty` to address this issue.

**Comment by jb2197** [comment]
Fixed in [py4jps==1.0.37](https://pypi.org/project/py4jps/1.0.37/), stable release in #854.
